opencti-platform/opencti-graphql/src/utils/stix-filtering/boolean-logic-engine.ts

```diff
@@ -0,0 +1,187 @@
+ * @param stixCandidates the values inside the DATA that we compare to the filter values; they are properly types
+ *                       We always assume an array of value(s) ; use toValidArray if the data is a single, nullable value.
+ */
+export const testGenericFilter = <T extends string | number | boolean>({ mode, operator }: FilterLogic, adaptedFilterValues: T[], stixCandidates: T[]) => {
+  // "(not) nil" or "(not) equal to nothing" is resolved the same way
+  if (operator === 'nil' || (operator === 'eq' && adaptedFilterValues.length === 0)) {
+    return stixCandidates.length === 0;
+  }
+  if (operator === 'not_nil' || (operator === 'not_eq' && adaptedFilterValues.length === 0)) {
+    return stixCandidates.length > 0;
+  }
+
+  // excluding the cases above, comparing to nothing is not supported and would never match
+  if (adaptedFilterValues.length === 0) {
+    return false;
+  }
+  if (mode === 'and') {
+    // we need to find all of them or none of them
+    return (operator === 'eq' && adaptedFilterValues.every((v) => stixCandidates.includes(v)))
+      || (operator === 'not_eq' && adaptedFilterValues.every((v) => !stixCandidates.includes(v)))
+
+      // In real cases, there is only 1 filter value with the next operators (not much sense otherwise)
+      || (operator === 'lt' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c < v)))
+      || (operator === 'lte' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c <= v)))
+      || (operator === 'gt' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c > v)))
+      || (operator === 'gte' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c >= v)));
+  }
+
+  if (mode === 'or') {
+    // we need to find one of them or at least one is not found
+    return (operator === 'eq' && adaptedFilterValues.some((v) => stixCandidates.includes(v)))
+      || (operator === 'not_eq' && adaptedFilterValues.some((v) => !stixCandidates.includes(v)))
+
+      // In real cases, there is only 1 filter value with the next operators (not much sense otherwise)
+      || (operator === 'lt' && adaptedFilterValues.some((v) => stixCandidates.some((c) => c < v)))
+      || (operator === 'lte' && adaptedFilterValues.some((v) => stixCandidates.some((c) => c <= v)))
+      || (operator === 'gt' && adaptedFilterValues.some((v) => stixCandidates.some((c) => c > v)))
+      || (operator === 'gte' && adaptedFilterValues.some((v) => stixCandidates.some((c) => c >= v)));
+  }
+
+  return false;
+};
+
+/**
+ * Implementation of testGenericFilter for string values.
+ * String comparison is insensitive to case.
+ */
+export const testStringFilter = (filter: FilterExcerpt, stixCandidates: string[]) => {
+  const filterValuesLowerCase = filter.values.map((v) => v.toLowerCase());
+  const stixValuesLowerCase = stixCandidates.map((v) => v.toLowerCase());
```

Comment:
Worth noting : we're case insensitive by default.

Reply:
I don't understand, why are we doing the mapping if it worth nothing ?

Reply:
worth noting. Like "worth to note" :)